### PR TITLE
[GNA] Support for cascade concat with non functional layers

### PR DESCRIPTION
**Pull request by admitriev-gna**
Support for cascade concat with non functional layers between concats.
Added tests for cascade concat

### CODE REVIEW

**Comment by admitriev-gna** [comment]
@dorloff, @esmirno  check please)

File: inference-engine/src/gna_plugin/gna_graph_compiler.cpp [hunk 1]
```diff
@@ -550,6 +550,23 @@ void GNAGraphCompiler::ConcatPrimitive(InferenceEngine::CNNLayerPtr layer) {
     auto& concatLayerInfo = concat_connection.find(concatLayer->name)->second;
     for (auto &&outLayer : concatLayer->outData.front()->getInputTo()) {
         if ( LayerInfo(outLayer.second).isConcat() ) {
+            auto concatCadidate = outLayer.second;
```
Comment (pavel-rodionov) [inline inference-engine/src/gna_plugin/gna_graph_compiler.cpp:553]: Typo here

File: inference-engine/tests/functional/plugin/shared/include/subgraph_tests/multiply_concat_reshape_reshape_concat.hpp [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+// SPDX-License-Identifier: Apache-2.0
+#pragma once
+
+#include <tuple>
+#include <string>
+#include <vector>
+#include <memory>
+#include "functional_test_utils/layer_test_utils.hpp"
+#include "ngraph_functions/builders.hpp"
+#include "ngraph_functions/utils/ngraph_helpers.hpp"
```
Comment (pavel-rodionov) [inline inference-engine/tests/functional/plugin/shared/include/subgraph_tests/multiply_concat_reshape_reshape_concat.hpp:11]: I'm sorry but do you actually use something form  `ngraph_helpers.hpp` here in this header?

File: inference-engine/tests/functional/plugin/shared/src/subgraph_tests/multiply_concat_reshape_reshape_concat.cpp [hunk 3]
```diff
@@ -0,0 +1,57 @@
+// Copyright (C) 2020 Intel Corporation
+// SPDX-License-Identifier: Apache-2.0
+//
+#include <tuple>
+#include <string>
+#include <vector>
+#include <memory>
+#include <debug.h>
```
Comment (pavel-rodionov) [inline inference-engine/tests/functional/plugin/shared/src/subgraph_tests/multiply_concat_reshape_reshape_concat.cpp:8]: Why do we need this?

File: inference-engine/tests/functional/plugin/shared/src/subgraph_tests/multiply_concat_reshape_reshape_concat.cpp [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+        auto squeeze = ngraph::builder::makeUnsqueeze(concat, ngraph::element::i64, std::vector<size_t>{0});
+        auto squeeze2 = ngraph::builder::makeSqueeze(squeeze, ngraph::element::i64, std::vector<size_t>{0});
+        auto const_concat2 = std::make_shared<ngraph::opset1::Constant>(ngPrc,
+                                                                       ngraph::Shape{squeeze2->get_shape()},
+                                                                       std::vector<size_t>{1});
+        auto concat2 = std::make_shared<ngraph::opset1::Concat>(ngraph::OutputVector{squeeze2->output(0),
+                                                                                    const_concat2->output(0)},
+                                                               0);
+        auto const_mul = ngraph::builder::makeConstant(ngPrc, ngraph::Shape{1}, std::vector<float>{-1.0f});
+        auto mult = std::make_shared<ngraph::opset1::Multiply>(input[0], const_mul);
```
Comment (pavel-rodionov) [inline inference-engine/tests/functional/plugin/shared/src/subgraph_tests/multiply_concat_reshape_reshape_concat.cpp:50]: For consistency I can suggest your to use `const_mul` and `mul` or `const_mult` and `mult`

File: inference-engine/tests/functional/plugin/shared/src/subgraph_tests/multiply_concat_reshape_reshape_concat.cpp [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+        function = std::make_shared<ngraph::Function>(mult, input, "concat_reshape_reshape_concat_mul");
+    }
+
+    TEST_P(ConcatReshapeReshapeConcatMul, CompareWithRefs){
```
Comment (esmirno) [inline inference-engine/tests/functional/plugin/shared/src/subgraph_tests/multiply_concat_reshape_reshape_concat.cpp:54]: why we are not creating unit tests at all? Functional test to me only should happen if fight with accuracy. And event than it is questionable.

File: inference-engine/src/gna_plugin/gna_graph_compiler.cpp [hunk 6]
```diff
@@ -550,6 +550,23 @@ void GNAGraphCompiler::ConcatPrimitive(InferenceEngine::CNNLayerPtr layer) {
...
+            if (LayerInfo(concatCadidate).isNonFunctional()) {
+                // searching for next concat
+                auto isNonFunctional = [](CNNLayerPtr l) {
+                    return LayerInfo(l).isNonFunctional();
+                };
+                if (!CNNNetHasNextLayerSkipCertain(concatCadidate, 0, 0, isNonFunctional)) {
+                    continue;
+                }
+                concatCadidate = CNNNetGetNextLayerSkipCertain(concatCadidate, 0, 0, isNonFunctional).first;
```
Comment (esmirno) [inline inference-engine/src/gna_plugin/gna_graph_compiler.cpp:562]: what if there will be non linear subgraph of reshapes? What functional layer will be ended up with?
Create a case concat->reshape->(out:0)affine, (out:1 )concat

File: inference-engine/src/gna_plugin/gna_graph_compiler.cpp [hunk 7]
```diff
@@ -534,9 +534,22 @@ void GNAGraphCompiler::ConcatPrimitive(InferenceEngine::CNNLayerPtr layer) {
 
     auto& concatLayerInfo = concat_connection.find(concatLayer->name)->second;
     for (auto &&outLayer : concatLayer->outData.front()->getInputTo()) {
-        if ( LayerInfo(outLayer.second).isConcat() ) {
-            connectOutput(layer, &concatLayerInfo.gna_ptr, concatLayerInfo.reserved_size);
+        auto concatCadidate = outLayer.second;
```
Comment (dorloff) [inline inference-engine/src/gna_plugin/gna_graph_compiler.cpp:537]: CanNdidate

File: inference-engine/src/gna_plugin/gna_graph_compiler.cpp [hunk 8]
```diff
@@ -534,9 +534,22 @@ void GNAGraphCompiler::ConcatPrimitive(InferenceEngine::CNNLayerPtr layer) {
...
+            };
+            if (!CNNNetHasNextLayerSkipCertain(concatCadidate, 0, 0, isNonFunctional)) {
+                continue;
+            }
+            concatCadidate = CNNNetGetNextLayerSkipCertain(concatCadidate, 0, 0, isNonFunctional).first;
+        }
+        if (!LayerInfo(concatCadidate).isConcat()) {
+            continue;
         }
+        gnalog() << "Cascaded concat connection found  from: " << layer->name << ", to: " << concatCadidate->name << std::endl;
```
Comment (dorloff) [inline inference-engine/src/gna_plugin/gna_graph_compiler.cpp:551]: Extra blank after "found"

File: inference-engine/src/gna_plugin/gna_graph_compiler.cpp [hunk 9]
```diff
@@ -557,21 +570,32 @@ void GNAGraphCompiler::ConcatPrimitive(InferenceEngine::CNNLayerPtr layer) {
         auto layerInfo = LayerInfo(concatParent);
         // auto layerInfo = LayerInfo(concatLayerInput->insData[it].lock()->getCreatorLayer().lock());
         if (layerInfo.isInput()) {
+            auto & bytesAllocated = inputDesc->bytes_allocated_for_input[((InferenceEngine::CNNLayerPtr)layerInfo)->name];
             if (concatLayerInfo.input_allocated) {
+//                if (0 != bytesAllocated) {
+//                    // we are remap input pointer to be not 64 bytes aligned which might be not possible in case
```
Comment (dorloff) [inline inference-engine/src/gna_plugin/gna_graph_compiler.cpp:576]: "We are remapping"

File: inference-engine/src/gna_plugin/gna_graph_compiler.cpp [hunk 10]
```diff
@@ -557,21 +570,32 @@ void GNAGraphCompiler::ConcatPrimitive(InferenceEngine::CNNLayerPtr layer) {
         auto layerInfo = LayerInfo(concatParent);
         // auto layerInfo = LayerInfo(concatLayerInput->insData[it].lock()->getCreatorLayer().lock());
         if (layerInfo.isInput()) {
+            auto & bytesAllocated = inputDesc->bytes_allocated_for_input[((InferenceEngine::CNNLayerPtr)layerInfo)->name];
             if (concatLayerInfo.input_allocated) {
+//                if (0 != bytesAllocated) {
```
Comment (dorloff) [inline inference-engine/src/gna_plugin/gna_graph_compiler.cpp:575]: Why are we adding commented code?

File: inference-engine/tests/functional/plugin/shared/src/subgraph_tests/cascade_concat.cpp [hunk 11]
```diff
@@ -0,0 +1,65 @@
...
+            results = ngraph::ResultVector{std::make_shared<ngraph::opset1::Result>(concat2),
+                                           std::make_shared<ngraph::opset1::Result>(mult)};
+        } else {
+            results = ngraph::ResultVector{std::make_shared<ngraph::opset1::Result>(concat2)};
+        }
+        function = std::make_shared<ngraph::Function>(results, input, "concat_reshape_reshape_concat_mul");
+    }
+
+    TEST_P(CascadeConcat, CompareWithRefs){
+    Run();
```
Comment (dorloff) [inline inference-engine/tests/functional/plugin/shared/src/subgraph_tests/cascade_concat.cpp:63]: Seems like indentation is wrong

File: inference-engine/tests/functional/plugin/shared/src/subgraph_tests/cascade_concat.cpp [hunk 12]
```diff
@@ -0,0 +1,65 @@
...
+                                           std::make_shared<ngraph::opset1::Result>(mult)};
+        } else {
+            results = ngraph::ResultVector{std::make_shared<ngraph::opset1::Result>(concat2)};
+        }
+        function = std::make_shared<ngraph::Function>(results, input, "concat_reshape_reshape_concat_mul");
+    }
+
+    TEST_P(CascadeConcat, CompareWithRefs){
+        Run();
+};
```
Comment (dorloff) [inline inference-engine/tests/functional/plugin/shared/src/subgraph_tests/cascade_concat.cpp:64]: The indentation is wrong here, too

**Pull request merged**